src/server/autocomplete/util.d

```diff
@@ -706,36 +666,69 @@ body
 	case tok!"[":
 		return i + 1;
 	case tok!")":
-		open = tok!")";
-		close = tok!"(";
-		goto skip;
+		beforeTokens.skipParenReverse!true(i, tok!")", tok!"(");
+		break;
 	case tok!"}":
-		open = tok!"}";
-		close = tok!"{";
-		goto skip;
+		beforeTokens.skipParenReverse!true(i, tok!"}", tok!"{");
+		break;
 	case tok!"]":
-		open = tok!"]";
-		close = tok!"[";
-	skip:
-		if (i == 0)
-			return size_t.max;
-		else
-			i--;
-		int depth = 1;
-		do
-		{
-			if (depth == 0 || i == 0)
-				break;
-			else
-				i--;
-			if (beforeTokens[i].type == open)
-				depth++;
-			else if (beforeTokens[i].type == close)
-				depth--;
-		} while (true);
+		beforeTokens.skipParenReverse!true(i, tok!"]", tok!"[");
 		break;
 	default:
 		return size_t.max;
 	}
-	return size_t.max;
+}
+
+/**
+ * Skips blocks of parentheses until the starting block has been closed
+ */
+void skipParen(T)(T tokenSlice, ref size_t i, IdType open, IdType close)
+{
+	if (i >= tokenSlice.length || tokenSlice.length <= 0)
+		return;
+	int depth = 1;
+	while (depth != 0 && i + 1 != tokenSlice.length)
+	{
+		i++;
+		if (tokenSlice[i].type == open)
+			depth++;
+		else if (tokenSlice[i].type == close)
+			depth--;
+	}
+}
+
+/**
+ * Skips blocks of parentheses in reverse until the starting block has been opened
+ */
+void skipParenReverse(bool before = false, T)(T beforeTokens, ref size_t i, IdType open, IdType close)
```

Comment:
Just a tiny nitpick here, rather than using a template here for "before", you could do something like this:

```D

// Don't think we need ref here for i either
size_t skipParenReverse(T)(T beforeTokens, /*ref*/ size_t i, IdType open, IdType close)
{
    // ...
}

size_t skipParenReverseBefore(T)(T beforeTokens, size_t i, IdType open, IdType close)
{
    i = skipParenReverse(beforeTokens, i, open, close);
    if (i != 0)
        i -= 1;

    return i;
}
```

I find it makes it easier to read, you don't have to go searching to find out what the "true" means. Other than that nitpick everything looks good.